SwiftCompilerSources/Sources/Optimizer/Utilities/Cloner.swift

```diff
@@ -0,0 +1,186 @@
+
+import OptimizerBridging
+import SIL
+import SILBridging
+
+/// Clones the initializer value of a GlobalVariable.
+///
+/// Used to transitively clone "constant" instructions, including their operands,
+/// from or to the static initializer value of a GlobalVariable.
+///
+struct Cloner<Context: MutatingContext> {
+  private var bridged: BridgedCloner
+  let context: Context
+
+  enum Target {
+    case function(Function)
+    case global(GlobalVariable)
+  }
+  let target: Target
+
+  init(cloneToGlobal: GlobalVariable, _ context: Context) {
+    self.bridged = BridgedCloner(cloneToGlobal.bridged, context.bridgedPassContext)
+    self.context = context
+    self.target = .global(cloneToGlobal)
+  }
+
+  init(cloneBefore inst: Instruction, _ context: Context) {
+    self.bridged = BridgedCloner(inst.bridged, context.bridgedPassContext)
+    self.context = context
+    self.target = .function(inst.parentFunction)
+  }
+
+  init(cloneToEmptyFunction: Function, _ context: Context) where Context == FunctionPassContext {
+    self.bridged = BridgedCloner(cloneToEmptyFunction.bridged, context.bridgedPassContext)
+    self.context = context
+    self.target = .function(cloneToEmptyFunction)
+  }
+
+  mutating func deinitialize() {
+    bridged.destroy(context.bridgedPassContext)
+  }
+
+  var targetFunction: Function {
+    guard case .function(let function) = target else {
+      fatalError("expected cloning into a function")
+    }
+    return function
+  }
+
+  mutating func clone(instruction: Instruction) -> Instruction {
+    let cloned = bridged.clone(instruction.bridged).instruction
+    if case .function = target {
+      context.notifyInstructionChanged(cloned)
+      context.notifyInstructionsChanged()
+    }
+    return cloned
+  }
+
+  /// Transitively clones `value` including its defining instruction's operands.
+  mutating func cloneRecursively(value: Value) -> Value {
+    if isCloned(value: value) {
+      return getClonedValue(of: value)
+    }
+    
+    if let beginAccess = value as? BeginAccessInst {
+      // Skip access instructions, which might be generated for UnsafePointer globals which point to other globals.
+      let clonedOperand = cloneRecursively(value: beginAccess.address)
+      bridged.recordFoldedValue(beginAccess.bridged, clonedOperand.bridged)
+      return clonedOperand
+    }
+
+    guard let inst = value.definingInstruction else {
+      fatalError("expected instruction to clone or already cloned value")
+    }
+
+    for op in inst.operands {
+      _ = cloneRecursively(value: op.value)
+    }
+
+    let cloned = clone(instruction: inst)
+    if let svi = cloned as? SingleValueInstruction {
+      return svi
+    } else if let originalMvi = value as? MultipleValueInstructionResult {
+      return cloned.results[originalMvi.index]
+    }
+    fatalError("unexpected instruction kind")
+  }
+
+  mutating func cloneUseDefChain(addr: Value, checkBase: (Value) -> Bool) -> Value? {
```

Comment:
`cloneUseDefChain` is doing pretty much the same thing as `cloneRecursively`. It would be great to implement it by using `cloneUseDefChain` with the appropriate closure.
Also, as this is specifically for address projections, let's call it `cloneAddressProjections`